README.md

```diff
@@ -99,27 +99,41 @@ Of course, there is overlap is the sets of problems the two approaches can uncov
+Conversely, consider a statement `S` that satisfies (`*`). Here is why it might make sense to remove `S`. Think of `S` as _shifting_ the set of valid environments, rather than constraining them. More precisely, if `S`'s weakest precondition `P` does not imply `Q`, and if `Q` is satisfiable, the there is an assignment to `P` and `Q`'s free variables that satisfies both `P` and `Q`. If such an assignment results from each environment in which `S` is actually executed, then the necessity of `S` is called into question.
 
-To be clear, as currently implemented, Necessist does not actually compute weakest preconditions. Rather, Necessist operates purely on syntax. Intuitively, given a statement `S`, Necessist makes a "best guess" as to what `S`'s semantics would be, and removes or ignores `S` based on that guess.
+The main utility of (`*`) is in helping to select the statements that Necessist ignores. That is, if we imagine a predicate transformer semantics for one of Necessist's supported languages, and a statement `S` in that language, we can ask: would `S` satisfy (`*`)? If not, then then Necessist should likely ignore `S`.
 
-Even though (`*`) is not applied formally, it provides a criterion to guide selection of the statements that Necessist ignores. That is, if we imagine a predicate transformer semantics for one of Necessist's supported languages, and a statement `S` in that language, we can ask: would `S` satisfy (`*`)? If not, then then Necessist should likely ignore `S`.
+But (`*`) has other nice consequences. For example, the rule that the last statement in a test should not be removed follows from (`*`). To see this, note the such a statement's postcondition `Q` is always `True`. Thus, if the statement doesn't change the context, then its weakest precondition necessarily implies `Q`.
 
-If you think one of Necessist's implicit rules violates this criterion, please open an issue!
+Having said all this, (`*`) doesn't quite capture what Necessist actually _does_. The next example is due to @2over12. Consider a statement `x.foo();` with a weakest precondition and postcondition as follows:
+
+```rust
+{ Q ^ x.y == x.z ^ x.y > 2}
+x.foo();
+{ Q }
+```
+
+Unless `x.foo` is ignored via a [configuration file], Necessist will remove `x.foo();`. It does so despite the fact the `x.foo();` does not change the context, and `Q ^ x.y == x.z ^ x.y > 2` implies `Q`.
+
+However, it's not entirely clear whether Necessist _should_ remove this statement. The condition `x.y == x.z ^ x.y > 2` could be the result of side effects that were unintentional. In that case, knowing whether the enclosing test passes without `x.foo();` could be useful.
```

Comment:
This is a bit imprecise to me imo. My major concern is in an SP calculus we could have the information:
```
{ Q ^ x.y == x.z ^ x.y > 2}
x.foo();
{ Q /\ x.y' == x.y +1 }
```
We are imagining here that foo checks some state on y and z then increments. But say there is no assert related to x.y' and x.y in the rest of the test then we still get a WP result:
```
{ Q ^ x.y == x.z ^ x.y > 2}
x.foo();
{ Q }
```
Since the post condition is essentially trivially satisfied wrt to Q. ie. my more major concern is that (*) drops unasserted effects, which is often the source of a buggy/under-constrained test. ie. in a very rough sense I think `verify_untrusted_callback_override_ok` is an example of a test that doesnt really uphold (*) since the post condition on the callback is essentially True

On the other hand, state being tested by asserts isnt a bad signal that that portion of the state is interesting. other statements that touch related/the same state are worth attempting to remove

 This is a bit pedantic tho, the point remains that there are probably actions that should be/are removed that may not reflect `(*)`

Reply:
Could you please give this another glance when you have a moment?